libtransmission/peer-mgr.cc

```diff
@@ -244,15 +288,28 @@ class tr_swarm
         TR_ASSERT(stats.peer_count == peerCount());
 
         delete peer;
+
+        if (was_incoming)
+        {
+            [[maybe_unused]] auto const port_empty = std::empty(peer_info->listen_port());
+            if (incoming_pool.erase(socket_address) != 0U)
+            {
+                TR_ASSERT(port_empty);
+            }
+        }
+        graveyard_pool.erase(socket_address);
     }
 
-    void removeAllPeers()
+    void remove_peer(tr_peerMsgs* peer)
     {
-        auto tmp = peers;
+        remove_peer(std::find(std::begin(peers), std::end(peers), peer));
+    }
 
-        for (auto* peer : tmp)
+    void remove_all_peers()
+    {
+        for (auto rit = std::rbegin(peers); rit != std::rend(peers); rit = std::rbegin(peers))
         {
-            removePeer(peer);
+            remove_peer(rit.base() - 1);
```

Comment:
These two variants of `remove_peer()` work but the iterator / reverse iteration is a little awkward.

Just brainstorming here to see what you think -- how about a function like `close_peer(tr_peerMsgs*)` that does everything in this function _except_ remove the peer from `swarm.peers`, and then 

```c++
remove_peer(tr_peerMsgs* peer)
{
    if (auto iter = std::find(std::begin(peers), std::end(peers), peer); iter != std::end(peers))
    {
        peers.erase(peer);
    }
    close_peer(peer);
}

remove_all_peers()
{
    for (auto* peer : peers)
    {
        close_peer(peer);
    }
    peers.clear();
}

Reply:
This does make it easier to read.

I would move `close_peer()` into the if condition in `remove_peer()` though.

Reply:
36015c0